src/compiler/checker.ts

```diff
@@ -22365,7 +22384,63 @@ namespace ts {
                 // succeed. That would be a pointless and confusing outcome.
                 if (matches || every(target.texts, s => s.length === 0)) {
                     for (let i = 0; i < types.length; i++) {
-                        inferFromTypes(matches ? matches[i] : neverType, types[i]);
+                        const source = matches ? matches[i] : neverType;
+                        const target = types[i];
+
+                        // If we are inferring from a string literal type to a type variable whose constraint includes one of the
+                        // allowed template literal placeholder types, infer from a literal type corresponding to the constraint.
+                        let sourceTypes: Type[] | undefined;
+                        if (source.flags & TypeFlags.StringLiteral && target.flags & TypeFlags.TypeVariable) {
+                            const inferenceContext = getInferenceInfoForType(target);
+                            const constraint = inferenceContext ? getConstraintOfTypeParameter(inferenceContext.typeParameter) : undefined;
+                            if (inferenceContext && constraint) {
+                                const str = (source as StringLiteralType).value;
+                                const constraintTypes = constraint.flags & TypeFlags.Union ? (constraint as UnionType).types : [constraint];
+                                for (const constraintType of constraintTypes) {
+                                    if (constraintType.flags & TypeFlags.StringLike) {

```

Comment:
All of these `if` statements could be turned into a single ternary operator that just computes which type to replace `constraintType` with, followed by a common `push` operation.